crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2330,10 +2296,62 @@ impl<'db> TypeInferenceBuilder<'db> {
                 );
                 self.diagnostics
                     .add_unresolved_module(import_from, *level, module);
-                Type::Unknown
+                self.add_unknown_declaration_with_binding(alias.into(), definition);
+                return;
+            }
+        };
+
+        let Some(module_ty) = self.module_ty_from_name(&module_name) else {
+            self.diagnostics
+                .add_unresolved_module(import_from, *level, module);
+            self.add_unknown_declaration_with_binding(alias.into(), definition);
+            return;
+        };
+
+        let ast::Alias {
+            range: _,
+            name,
+            asname: _,
+        } = alias;
+
+        // Check if the symbol being imported is a submodule.  (This won't get handled by the
+        // `Type::member` call below because the semantic index doesn't know whether this is a
+        // submodule being imported or an attribute being loaded, and so won't add it to the file's
+        // `imported_modules` set.)
```

Comment:
What if we did record each `from` import as a potential submodule import in semantic indexing? I think this might work, because when it isn't a submodule, we'd just get a little further into the submodule stanza in `Type::member` before finding we can't resolve the "submodule" and giving up and continuing to a regular module-namespace lookup. Then we wouldn't need to have this stanza both places, but I think also it would make code like this work, which doesn't in the current PR (this is the new code for `bar.py` in the test fixed by this PR):

```py
from . import foo
import pkg

reveal_type(pkg.foo.X)
```

That said: I'm not sure we need to make that code work. It's pretty subtle why it works at runtime, and (just as in the case without the `from . import foo` line) there's no reason you can't change `import pkg` to `import pkg.foo` to make it work explicitly.

And adding every `from` import as a speculative submodule import is going to increase the size of the data we have to store in semantic index, potentially a lot (especially for codebases that mostly use `from` imports of individual objects, which is definitely a style preferred some places).

So I think I'm happy going with this approach for now.

Reply:
> And adding every `from` import as a speculative submodule import is going to increase the size of the data we have to store in semantic index

I was worried less about the size of the data, since there are compressible representations if that turns out to be a problem.  My concern was more about execution time, since we would attempt a submodule resolution every time we look up a member of a module.

Though as I talk through this, I realize it wouldn't be for _every_ module member lookup — just the ones that are involved in a `from...import`.  Which would be a weird pattern to follow:

```py
import unittest
from unittest import mock
reveal_type(unittest.mock)
```

So maybe it wouldn't be so bad?  I do like the idea of these imports being handled by the existing code from #14946, and not needing the additional check.

Reply:
Ah, but tracking this in the semantic index means that we'd need to resolve relative import paths, which means we'd need to know (when building the semantic index) whether the current file is a module or a package.

Reply:
Makes sense! I don't have strong feelings about which way we go here, happy to go with whatever makes more sense to you. I think the most valuable thing is that we record in some form our conclusions from the time we've spent thinking about the two options here :) (The code example I gave above may be worth adding as a test either way.)

Reply:
> we'd need to resolve relative import paths, which means we'd need to know (when building the semantic index) whether the current file is a module or a package.

Ah! Yeah, that's a problem; we can't have dependencies on information external to the file in semantic indexing.

Reply:
Added that test case and some additional commentary